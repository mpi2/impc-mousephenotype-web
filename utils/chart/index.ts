
export const chartColors = [
  'rgba(9, 120, 161, 1)',
  'rgba(255, 201, 67, 1)',
  'rgba(239, 123, 11, 1)',
  'rgba(119, 119, 119, 1)',
  'rgba(36, 139, 75, 1)',
  'rgba(238, 238, 180, 1)',
  'rgba(191, 75, 50, 1)',
  'rgba(191, 151, 50, 1)',
  'rgba(239, 123, 11, 1)',
  'rgba(247, 157, 70, 1)',
  'rgba(247, 181, 117, 1)',
  'rgba(191, 75, 50, 1)',
  'rgba(151, 51, 51, 1)',
];