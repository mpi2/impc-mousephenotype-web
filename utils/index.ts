import _ from "lodash";

export const formatBodySystems = (systems: string[] | string = []) => {
  return _.capitalize(
    (typeof systems === "string" ? systems : systems.join(", "))
      .replace(/ phenotype/g, "")
      .replace(/\//g, " / ")
  );
};

export const formatAlleleSymbol = (allele: string) => {
  return allele.slice(0, allele.length - 1).split("<");
};
