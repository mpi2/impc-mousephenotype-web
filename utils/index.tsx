import React from "react";
import _ from "lodash";

export const formatBodySystems = (systems: string[] | string = []) => {
  return _.capitalize(
    (typeof systems === "string" ? systems : systems.join(", "))
      .replace(/ phenotype/g, "")
      .replace(/\//g, " / ")
  );
};

export const formatAlleleSymbol = (allele: string) => {
  return allele.slice(0, allele.length - 1).split("<");
};

export const formatPValue = (pValue: number) => {
  const pValueArray = Number.parseFloat(String(pValue))
    .toExponential(2)
    .split("e");
  return (
    <>
      {pValueArray[0]}x10<sup>{pValueArray[1].replace("+", "")}</sup>
    </>
  );
};

export const formatESCellName = (src: string) => {
  const [name, superscript, end] = src.split(/<|>/);
  return (
    <>
      {name}
      <sup>{superscript}</sup>
      {end}
    </>
  );
};

const fetchCache = {};
export { fetchCache };

export const toSentenceCase = (camelCase) => {
  if (camelCase) {
    const result = camelCase.replace(/([A-Z])/g, " $1");
    return result[0].toUpperCase() + result.substring(1).toLowerCase();
  }
  return "";
};
