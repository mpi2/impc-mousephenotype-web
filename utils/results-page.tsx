import DOMPurify from "dompurify";
import ReactHtmlParser from "react-html-parser";
export const surroundWithMarkEl = (text: string, query: string) => {
  const matchingSubsIndex = text.toLowerCase().indexOf(query.toLowerCase());
  const originalTextMatchedQuery =
    matchingSubsIndex !== -1
      ? DOMPurify.sanitize(
          text.slice(matchingSubsIndex, matchingSubsIndex + query.length)
        )
      : "";
  if (!!query) {
    return ReactHtmlParser(
      text.replaceAll(
        originalTextMatchedQuery,
        `<mark style="padding: 0;">${originalTextMatchedQuery}</mark>`
      )
    );
  }
  return text;
};
