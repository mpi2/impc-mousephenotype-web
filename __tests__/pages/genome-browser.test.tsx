import "@testing-library/jest-dom";
import { renderWithClient } from "../utils";
import GenomeBrowserPage from "@/app/genes/[pid]/genome-browser/genome-browser-page";

const ascc2Summary = {
  id: "MGI:1922702",
  mgiGeneAccessionId: "MGI:1922702",
  geneName: "activating signal cointegrator 1 complex subunit 2",
  geneSymbol: "Ascc2",
  markerType: "Gene",
  synonyms: ["1700011I11Rik", "ASC1p100", "2610034L15Rik"],
  humanGeneSymbols: ["ASCC2"],
  humanSymbolSynonyms: ["FLJ21588", "ASC1p100", "DKFZp586O0223"],
  productionCentres: ["BCM"],
  phenotypingCentres: ["BCM"],
  alleleNames: [
    "targeted mutation 1a, Wellcome Trust Sanger Institute",
    "endonuclease-mediated mutation 11, GemPharmatech Co., Ltd",
    "endonuclease-mediated mutation 1, GemPharmatech Co., Ltd",
    "targeted mutation 1b, Wellcome Trust Sanger Institute",
  ],
  isUmassGene: false,
  isIdgGene: false,
  analysisDownloadUrl: null,
  embryoAnalysisViewUrl: null,
  centre: null,
  colonyId: null,
  hasAutomatedAnalysis: null,
  mgi: null,
  proceduresParameters: null,
  url: null,
  embryoAnalysisViewName: null,
  embryoModalities: null,
  hasEmbryoImagingData: false,
  chrName: "11",
  seqRegionStart: "4587747",
  seqRegionEnd: "4635699",
  chrStrand: "+",
  seqRegionId: "11",
  entrezgeneId: null,
  ensemblGeneIds: null,
  ccdsIds: null,
  ncbiId: null,
  assignmentStatus: "Selected for production",
  nullAlleleProductionStatus: "Genotype confirmed mice",
  conditionalAlleleProductionStatus: "Genotype confirmed mice",
  crisprAlleleProductionStatus: null,
  crisprConditionalAlleleProductionStatus: null,
  esCellProductionStatus: "ES Cells Produced",
  mouseProductionStatus: "Mice Produced",
  phenotypingStatus: "Phenotyping All Data Sent",
  phenotypeStatus: "Phenotyping data available",
  phenotypingDataAvailable: true,
  significantTopLevelPhenotypes: [
    "homeostasis/metabolism phenotype",
    "immune system phenotype",
    "growth/size/body region phenotype",
    "hematopoietic system phenotype",
    "behavior/neurological phenotype",
    "mortality/aging",
    "cardiovascular system phenotype",
  ],
  notSignificantTopLevelPhenotypes: [
    "reproductive system phenotype",
    "integument phenotype",
    "adipose tissue phenotype",
    "muscle phenotype",
    "hearing/vestibular/ear phenotype",
    "craniofacial phenotype",
    "pigmentation phenotype",
    "embryo phenotype",
    "limbs/digits/tail phenotype",
    "nervous system phenotype",
    "vision/eye phenotype",
    "respiratory system phenotype",
    "liver/biliary system phenotype",
    "skeleton phenotype",
  ],
  significantPhenotypesCount: 7,
  adultExpressionObservationsCount: 49,
  embryoExpressionObservationsCount: 62,
  associatedDiseasesCount: 0,
  hasLacZData: true,
  hasImagingData: true,
  hasHistopathologyData: true,
  hasViabilityData: true,
  hasBodyWeightData: true,
};

jest.mock("next/navigation", () => ({
  useRouter: jest.fn(),
  useSearchParams: jest.fn().mockImplementation(() => new URLSearchParams()),
  usePathname: jest.fn(),
}));

jest.mock("igv/dist/igv.esm", () => {
  return {
    __esModule: true,
    default: {
      createBrowser: jest.fn().mockImplementation(() =>
        Promise.resolve({
          search: jest.fn(),
        }),
      ),
    },
  };
});

describe("Genome Browser page", () => {
  it("renders properly", async () => {
    const { container } = renderWithClient(
      <GenomeBrowserPage
        mgiGeneAccessionId={ascc2Summary.mgiGeneAccessionId}
        geneSummary={ascc2Summary}
      />,
    );
    expect(container).toMatchSnapshot();
  });
});
