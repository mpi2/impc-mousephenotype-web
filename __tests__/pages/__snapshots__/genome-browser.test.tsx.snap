// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Genome Browser page renders properly 1`] = `
<div>
  <div
    class="banner"
  >
    <div
      class="pb-4 pt-5 container container"
    >
      <div
        class="col-12 col-md-8 ps-4 pe-4"
      >
        <div
          class="tabs"
        >
          <a
            class="tab__active"
            href="/search"
          >
            Genes
          </a>
          <a
            class="tab"
            href="/search?type=pheno"
          >
            Phenotypes
          </a>
          <a
            class="tab"
            href="https://dev.mousephenotype.org/?s="
          >
            Help, news, blog
          </a>
        </div>
        <div
          class="inputCont"
        >
          <input
            class="input"
            placeholder="Search for a gene..."
            title="main search box"
            type="text"
            value=""
          />
          <button
            aria-describedby="svg-inline--fa-title-search-icon"
            class="searchBtn"
          >
            <svg
              aria-labelledby="svg-inline--fa-title-search-icon"
              class="svg-inline--fa fa-magnifying-glass"
              data-icon="magnifying-glass"
              data-prefix="fas"
              role="img"
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <title
                id="svg-inline--fa-title-search-icon"
              >
                Search button
              </title>
              <path
                d="M416 208c0 45.9-14.9 88.3-40 122.7L502.6 457.4c12.5 12.5 12.5 32.8 0 45.3s-32.8 12.5-45.3 0L330.7 376c-34.4 25.2-76.8 40-122.7 40C93.1 416 0 322.9 0 208S93.1 0 208 0S416 93.1 416 208zM208 352a144 144 0 1 0 0-288 144 144 0 1 0 0 288z"
                fill="currentColor"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    class="page container"
  >
    <div
      class="card"
    >
      <div
        class="subheading mb-3"
      >
        <span
          class="subheadingSection primary"
          data-testid="back-to-gene-page-link"
        >
          <a
            class="mb-3 link"
            href="/genes/MGI:1922702"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-arrow-left"
              data-icon="arrow-left"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M9.4 233.4c-12.5 12.5-12.5 32.8 0 45.3l160 160c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L109.2 288 416 288c17.7 0 32-14.3 32-32s-14.3-32-32-32l-306.7 0L214.6 118.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0l-160 160z"
                fill="currentColor"
              />
            </svg>
            Â  Go Back to
             
            <i>
              <span>
                Ascc2
              </span>
            </i>
          </a>
        </span>
      </div>
      <div
        class="container"
        style="padding: 0px;"
      >
        <div
          class="row"
        >
          <div
            class="col"
          >
            <div
              class="mb-3 mt-0"
              style="display: flex; align-items: center;"
            >
              <h2
                style="margin: 0px;"
              >
                Genome browser
              </h2>
              <a
                aria-label="Genome browser documentation"
                class="btn"
                href="https://dev.mousephenotype.org/help/data-visualization/genome-browser/"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-circle-question fa-xl"
                  data-icon="circle-question"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M256 512A256 256 0 1 0 256 0a256 256 0 1 0 0 512zM169.8 165.3c7.9-22.3 29.1-37.3 52.8-37.3h58.3c34.9 0 63.1 28.3 63.1 63.1c0 22.6-12.1 43.5-31.7 54.8L280 264.4c-.2 13-10.9 23.6-24 23.6c-13.3 0-24-10.7-24-24V250.5c0-8.6 4.6-16.5 12.1-20.8l44.3-25.4c4.7-2.7 7.6-7.7 7.6-13.1c0-8.4-6.8-15.1-15.1-15.1H222.6c-3.4 0-6.4 2.1-7.5 5.3l-.4 1.2c-4.4 12.5-18.2 19-30.6 14.6s-19-18.2-14.6-30.6l.4-1.2zM224 352a32 32 0 1 1 64 0 32 32 0 1 1 -64 0z"
                    fill="currentColor"
                  />
                </svg>
              </a>
            </div>
          </div>
          <div
            class="resetBtnContainer col"
          >
            <button
              class="btn impc-secondary-button small"
            >
              Reset view
            </button>
          </div>
        </div>
        <div
          class="row"
          style="position: relative;"
        >
          <div
            class="col"
          >
            <div
              class="controlsContainer"
            >
              <div>
                <label
                  class="d-inline-block fst-italic me-3 mb-0 form-label"
                >
                  Gene annotations:
                </label>
                <div
                  class="mb-0 form-check form-check-inline"
                >
                  <input
                    checked=""
                    class="form-check-input"
                    type="checkbox"
                  />
                  <label
                    class="form-check-label"
                    title=""
                  >
                    RefSeq
                  </label>
                </div>
                <div
                  class="mb-0 form-check form-check-inline"
                >
                  <input
                    class="form-check-input"
                    id="gencode-complete"
                    type="checkbox"
                  />
                  <label
                    class="form-check-label"
                    for="gencode-complete"
                    title=""
                  >
                    GENCODE M37 (complete)
                  </label>
                </div>
                <div
                  class="mb-0 form-check form-check-inline"
                >
                  <input
                    class="form-check-input"
                    id="gencode-basic"
                    type="checkbox"
                  />
                  <label
                    class="form-check-label"
                    for="gencode-basic"
                  >
                    GENCODE M37 basic annotations
                  </label>
                </div>
              </div>
            </div>
            <hr />
            <div
              class="controlsContainer"
            >
              <div>
                <label
                  class="d-inline-block fst-italic me-3 mb-0 form-label"
                >
                  Protein annotations:
                </label>
                <div
                  class="mb-0 form-check form-check-inline"
                >
                  <input
                    class="form-check-input"
                    id="swissprot-annotations"
                    type="checkbox"
                  />
                  <label
                    class="form-check-label"
                    for="swissprot-annotations"
                    title=""
                  >
                    UniProt SwissProt Protein Annotations
                  </label>
                </div>
                <div
                  class="mb-0 form-check form-check-inline"
                >
                  <input
                    class="form-check-input"
                    id="trembl-annotations"
                    type="checkbox"
                  />
                  <label
                    class="form-check-label"
                    for="trembl-annotations"
                    title=""
                  >
                    TrEMBL Protein Annotations
                  </label>
                </div>
              </div>
            </div>
            <hr />
            <div
              class="controlsContainer"
            >
              <div>
                <label
                  class="d-inline-block fst-italic me-3 mb-0 form-label"
                >
                  Mouse products:
                </label>
                <div
                  class="mb-0 form-check form-check-inline"
                >
                  <input
                    class="form-check-input"
                    id="crispr-guides"
                    type="checkbox"
                  />
                  <label
                    class="form-check-label"
                    for="crispr-guides"
                    title=""
                  >
                    CRISPR guides
                  </label>
                </div>
                <div
                  class="mb-0 form-check form-check-inline"
                >
                  <input
                    class="form-check-input"
                    id="crispr-aligned-FASTA"
                    type="checkbox"
                  />
                  <label
                    class="form-check-label"
                    for="crispr-aligned-FASTA"
                    title=""
                  >
                    CRISPR Aligned FASTA
                  </label>
                </div>
                <div
                  class="mb-0 form-check form-check-inline"
                >
                  <input
                    class="form-check-input"
                    id="crispr-deletions"
                    type="checkbox"
                  />
                  <label
                    class="form-check-label"
                    for="crispr-deletions"
                    title=""
                  >
                    CRISPR deletions
                  </label>
                </div>
                <div
                  class="mb-0 form-check form-check-inline"
                >
                  <input
                    class="form-check-input"
                    id="es-cell-mice"
                    type="checkbox"
                  />
                  <label
                    class="form-check-label"
                    for="es-cell-mice"
                    title=""
                  >
                    ES Cell Mice
                  </label>
                </div>
              </div>
            </div>
            <hr />
            <div
              class="controlsContainer"
            >
              <div>
                <label
                  class="d-inline-block me-3 mb-0 fst-italic form-label"
                >
                  ES Cell products:
                </label>
                <div
                  class="mb-0 form-check form-check-inline"
                >
                  <input
                    class="form-check-input"
                    id="es-cells"
                    type="checkbox"
                  />
                  <label
                    class="form-check-label"
                    for="es-cells"
                    title=""
                  >
                    ES Cells
                  </label>
                </div>
                <div
                  class="mb-0 form-check form-check-inline"
                >
                  <input
                    class="form-check-input"
                    id="ikmc-alleles"
                    type="checkbox"
                  />
                  <label
                    class="form-check-label"
                    for="ikmc-alleles"
                    title=""
                  >
                    IKMC alleles
                  </label>
                </div>
                <div
                  class="mb-0 form-check form-check-inline"
                >
                  <input
                    class="form-check-input"
                    id="targeting-vectors"
                    type="checkbox"
                  />
                  <label
                    class="form-check-label"
                    for="targeting-vectors"
                    title=""
                  >
                    Targeting vectors
                  </label>
                </div>
              </div>
            </div>
          </div>
          <div
            class="overlay active"
          >
            <div
              class="spinner-border text-primary"
            />
          </div>
        </div>
      </div>
      <div
        id="igv-container"
      />
    </div>
  </div>
</div>
`;
